### PR TITLE
add addon namespace validations

### PR DESCRIPTION
**Pull request by ashishmax31**
Signed-off-by: ashish <asnaraya@redhat.com>

### CODE REVIEW

**Comment by yashvardhan-kukreja** [comment]
can you refactor AM0004 to AM0008?

AM0004 is currently corresponding to `icon` validators - https://github.com/mt-sre/addon-metadata-operator/wiki/AM0004

File: pkg/validators/am0004_namespace.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		}
+		matched := namespaceRegex.MatchString(namespace)
+		if !matched {
+			failedNamespaces = append(failedNamespaces, namespace)
+		}
+	}
+	if len(failedNamespaces) == 0 {
+		return true, []string{}
+	}
+	return false, failedNamespaces
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:65]:
```suggestion
	return len(failedNamespaces) == 0, failedNamespaces
```

File: pkg/validators/am0004_namespace.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	if includes(targetNamespace, exceptionList) {
+		return true
+	}
+
+	for _, namespace := range namespaceList {
+		if namespace == targetNamespace {
+			return true
+		}
+	}
+	return false
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:78]:
```suggestion
        return includes(targetNamespace, namespaceList)
```

Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:78]:
```suggestion
	return includes(targetNamespace, exceptionList) || includes(targetNamespace, namespaceList)
```

File: pkg/validators/am0004_namespace.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func init() {
+	Registry.Add(AM0004)
+}
+
+func ValidateNamespace(metabundle utils.MetaBundle) (bool, string, error) {
+	targetNamespace := metabundle.AddonMeta.TargetNamespace
+	namespaceList := metabundle.AddonMeta.Namespaces
+	valid := validateNamespacePresence(targetNamespace, namespaceList, namespacePresenceExceptions)
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:39]: > we can get rid of validateNamespacePresence(...) function too as it's a single-liner anyway.

https://github.com/mt-sre/addon-metadata-operator/pull/44/files#r776232097

Thoughts?

File: pkg/validators/am0004_namespace.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+var AM0004 = utils.Validator{
+	Code:        "AM0004",
+	Name:        "ensure_namespace",
+	Description: "Ensure that the target namespace is listed in the set of channels listed",
+	Runner:      ValidateNamespace,
+}
+
+func init() {
+	Registry.Add(AM0004)
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:33]: s/AM0004/AM0008

File: pkg/validators/am0004_namespace.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+var namespacePresenceExceptions = []string{
+	"openshift-logging",
+}
+
+var namespaceRegex = regexp.MustCompile(`^redhat-.*$`)
+
+var AM0004 = utils.Validator{
+	Code:        "AM0004",
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:26]: s/AM0004/AM0008

File: pkg/validators/test_am0004.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+				TargetNamespace: "redhat-random-operator",
+				Namespaces: []string{
+					"redhat-random-operator",
+				},
+			},
+		},
+	}
+}
+
+func (val TestAM0004) FailingCandidates() []utils.MetaBundle {
```
Comment (yashvardhan-kukreja) [inline pkg/validators/test_am0004.go:36]: s/AM0004/AM0008

File: pkg/validators/test_am0004.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	return []utils.MetaBundle{
+		{
+			AddonMeta: &v1alpha1.AddonMetadataSpec{
+				ID:              "random-operator",
+				TargetNamespace: "redhat-random-operator",
+				Namespaces: []string{
+					"redhat-random-operator",
+				},
+			},
+		},
```
Comment (yashvardhan-kukreja) [inline pkg/validators/test_am0004.go:32]: can we attach `testutils.DefaultSucceedingCandidates()` as well to the returning slice?

Ref: https://github.com/mt-sre/addon-metadata-operator/blob/57f68f5398eef130f5416d4719eb9106f296da0f/pkg/validators/test_am0006.go#L25-L49

File: pkg/validators/am0004_namespace.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	targetNamespace := metabundle.AddonMeta.TargetNamespace
+	namespaceList := metabundle.AddonMeta.Namespaces
+	valid := validateNamespacePresence(targetNamespace, namespaceList, namespacePresenceExceptions)
+	if !valid {
+		return false, "Target namespace is not in the list of supplied namespaces", nil
+	}
+
+	allValid, failedNamespaces := validateNamespaceRegex(namespaceList, namespaceRegexExceptions)
+	if !allValid {
+		return false, fmt.Sprintf("Some namespaces doesn't start with 'redhat-*' %v", failedNamespaces), nil
```
Comment (yashvardhan-kukreja) [inline pkg/validators/am0004_namespace.go:46]:
```suggestion
		return false, fmt.Sprintf("Some namespaces don't start with 'redhat-*' %v", failedNamespaces), nil
```
minor nit

**Comment by yashvardhan-kukreja** [comment]
Do you mind creating a wiki page dedicated to this validator as well? 

https://github.com/mt-sre/addon-metadata-operator/wiki/AM0008

**Comment by sugarraysam** [comment]
Good work. After changes, let's make sure we build and run this validator against addons in `managed-tenants`. As per the SOP: https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/MT-SRE/sops/AMO-validators.md#integrate-a-validator-with-the-managed-tenants-gitops-tooling

After integration in `managed-tenants`, we will be able to deprecate `tasks/check/09_test_namespaces.py`. Maybe move it to `tasks/archived/09_test_namespaces.py` so it doesn't run in CI twice.

WDYT?

File: pkg/validators/am0004_namespace.go (hunk 1 above)
Comment (ashishmax31) [reply pkg/validators/am0004_namespace.go:65]: Man i get lost go's verbosity, thanks :)

File: pkg/validators/am0004_namespace.go (hunk 2 above)
Comment (ashishmax31) [reply pkg/validators/am0004_namespace.go:78]: @yashvardhan-kukreja  I think we can keep the function, it adds more clarity to someone whose reading the code later. 
I think `return includes(targetNamespace, exceptionList) || includes(targetNamespace, namespaceList)` this refactor is a little cryptic.
I think this is more readable, Wdyt?:
```
if includes(targetNamespace, exceptionList) {
     return true
}

if includes(targetNamespace, namespaceList){
     return true
}
return false
```

**Comment by ashishmax31** [comment]
@yashvardhan-kukreja Yep, i'll create the wiki page once this PR is merged.

**Comment by ashishmax31** [comment]
@sugarraysam Yep that sounds good.

File: pkg/validators/am0004_namespace.go (hunk 2 above)
Comment (yashvardhan-kukreja) [reply pkg/validators/am0004_namespace.go]: I am a big proponent of cleaner code so I might be having a personal bias towards `return includes(targetNamespace, exceptionList) || includes(targetNamespace, namespaceList)`
or
```
// If targetNamespace is in the exceptionlist, return true
if includes(targetNamespace, exceptionList) {
     return true
}
// If targetNamespace is in the namespaceList, return true
returns includes(targetNamespace, namespaceList)
```

But yeah, it's not a hard requirement from my end coz it's not creating any performance-oriented difference.
So, feel free to move ahead with anything and resolve this thread :)
Thanks!

File: pkg/validators/am0004_namespace.go (hunk 1 above)
Comment (yashvardhan-kukreja) [reply pkg/validators/am0004_namespace.go]: Haha, no worries :)

File: pkg/validators/am0004_namespace.go (hunk 2 above)
Comment (ashishmax31) [reply pkg/validators/am0004_namespace.go]: @yashvardhan-kukreja 
Lets do `includes(targetNamespace, exceptionList) || includes(targetNamespace, namespaceList)`  :)

**Comment by ashishmax31** [comment]
@yashvardhan-kukreja PTAL

**Review by yashvardhan-kukreja — APPROVED** [review]
lgtm

**Pull request merged**